### PR TITLE
[SPARK-40819][SQL][3.3] Timestamp nanos behaviour regression

### PR DESCRIPTION
**Pull request by awdavidson**
As per @HyukjinKwon request on https://github.com/apache/spark/pull/38312 to backport fix into 3.3
### What changes were proposed in this pull request?

Handle `TimeUnit.NANOS` for parquet `Timestamps` addressing a regression in behaviour since 3.2

### Why are the changes needed?

Since version 3.2 reading parquet files that contain attributes with type `TIMESTAMP(NANOS,true)` is not possible as ParquetSchemaConverter returns
```
Caused by: org.apache.spark.sql.AnalysisException: Illegal Parquet type: INT64 (TIMESTAMP(NANOS,true))
```
https://issues.apache.org/jira/browse/SPARK-34661 introduced a change matching on the `LogicalTypeAnnotation` which only covers Timestamp cases for `TimeUnit.MILLIS` and `TimeUnit.MICROS` meaning `TimeUnit.NANOS` would return `illegalType()`

Prior to 3.2 the matching used the `originalType` which for `TIMESTAMP(NANOS,true)` return `null` and therefore resulted to a `LongType`, the change proposed is too consider `TimeUnit.NANOS` and return `LongType` making behaviour the same as before.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added unit test covering this scenario.
Internally deployed to read parquet files that contain `TIMESTAMP(NANOS,true)`

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Mind retriggering https://github.com/awdavidson/spark/actions/runs/4103316117/jobs/7077359507?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs).

cc @viirya since he is the release manager of Apache Spark 3.3.2.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3467]: Seems this is worth documenting it in migration guide?

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (awdavidson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3467]: I will look at adding to the guide

**Comment by HyukjinKwon** [comment]
Merged to branch-3.3.

**Pull request closed without merging**
